llvm/lib/SYCLLowerIR/SYCLConditionalCallOnDevice.cpp

```diff
@@ -0,0 +1,120 @@
+                    cl::init(""));
+
+PreservedAnalyses
+SYCLConditionalCallOnDevicePass::run(Module &M, ModuleAnalysisManager &) {
+  // find call_if_on_device_conditionally function
+  SmallVector<Function *, 4> FCallers;
+  for (Function &F : M.functions()) {
+    if (F.isDeclaration())
+      continue;
+
+    if (CallingConv::SPIR_KERNEL == F.getCallingConv())
+      continue;
+
+    if (F.getName().contains("call_if_on_device_conditionally") &&
+        !F.getName().contains("call_if_on_device_conditionally_helper"))
+      FCallers.push_back(&F);
+  }
+
+  int FCallerIndex = 1;
+  for (Function *FCaller : FCallers) {
+    // Find call to @CallableXXX in call_if_on_device_conditionally function
+    // (FAction). FAction should be a literal (i.e. not a pointer). The
+    // structure of the header file ensures that there is exactly one such
+    // instruction.
+    Function *FAction = nullptr;
+    for (Instruction &I : instructions(FCaller)) {
+      if (auto *CI = dyn_cast<CallInst>(&I)) {
+        FAction = CI->getCalledFunction();
+        break;
+      }
+    }
+
+    if (!FAction)
+      continue;
+
+    // Create a new function type with an additional function pointer argument
+    SmallVector<Type *, 4> NewParamTypes;
+    Type *FActionType = FAction->getType();
+    NewParamTypes.push_back(
+        PointerType::getUnqual(FActionType)); // Add function pointer to FAction
+    FunctionType *OldFCallerType = FCaller->getFunctionType();
+    for (Type *Ty : OldFCallerType->params())
+      NewParamTypes.push_back(Ty);
+
+    FunctionType *NewFCallerType =
+        FunctionType::get(OldFCallerType->getReturnType(), NewParamTypes,
+                          OldFCallerType->isVarArg());
+
+    // Create a new function with the updated type and rename it to
+    // call_if_on_device_conditionally_GUID_N
+    if (!UniquePrefixOpt.empty())
+      UniquePrefix = UniquePrefixOpt;
+    Twine NewFCallerName = Twine(FCaller->getName()) + "_" + UniquePrefix +
+                           "_" + Twine(FCallerIndex);
+    // Also change to external linkage
+    Function *NewFCaller = Function::Create(
+        NewFCallerType, Function::ExternalLinkage, NewFCallerName, &M);
+
+    NewFCaller->setCallingConv(FCaller->getCallingConv());
+
+    // Replace all calls to the old function with the new one
+    for (auto &U : FCaller->uses()) {
+      if (auto *Call = dyn_cast<CallInst>(U.getUser())) {
+        SmallVector<Value *, 4> Args;
+        // Add the function pointer as the first argument
+        Args.push_back(FAction);
+        for (unsigned i = 0; i < Call->arg_size(); ++i)
+          Args.push_back(Call->getArgOperand(i));
+
+        // Create the new call instruction
+        auto *NewCall =
+            CallInst::Create(NewFCaller, Args, /*	NameStr = */ "", Call);
+        NewCall->setCallingConv(Call->getCallingConv());
+        NewCall->setDebugLoc(Call->getDebugLoc());
+
+        // Replace the old call with the new call
```

Comment:
I'm a bit concerned about doing removals within this loop. You are modifying list of function uses while iterating over them and that's usually not a good idea, because it could cause iterator invalidation, or you may replace a call you have just inserted. A safer option would be to store a map `from -> to` and do replace + remove in a separate loop over that map.

Reply:
Good point, iterators here may be not robust, fixed

Reply:
Perhaps you forgot to push a commit? Because I don't see any changes

Reply:
Oh, I fixed the similar thing above, and didn't fix that one, applied now, thanks